firmware-management/mbl-watchdog-init/mbl-watchdog-init.c

```diff
@@ -0,0 +1,210 @@
+        case WDIOF_EXTERN2:
+            log_info("The last reboot was because external monitoring relay/source 2 was triggered.");
+        case WDIOF_POWERUNDER:
+            log_info("The last reboot was due to the machine showing an undervoltage status.");
+        case WDIOF_POWEROVER:
+            log_info("The last reboot was due to the machine showing an overvoltage status.");
+    }
+}
+
+
+int get_last_boot_status(const int watchdog_fd)
+{
+    int flags;
+
+    const int gs_ret = ioctl(watchdog_fd, WDIOC_GETBOOTSTATUS, &flags);
+    if (is_err(gs_ret))
+    {
+        return gs_ret;
+    }
+
+    print_last_boot_reason(flags);
+
+    return 0;
+}
+
+
+//============================================================================
+// Convert a string containing only digits to an integer. Exits on errors.
+int numeric_string_to_int(const char* const str)
+{
+    errno = 0;
+    char *endptr;
+    const long int res = strtol(str, &endptr, 0);
+
+    if ((errno == ERANGE && (res == LONG_MAX || res == LONG_MIN))
+           || (errno != 0 && res == 0))
+    {
+        log_error("strtol failed");
+        exit(EXIT_FAILURE);
+    }
+
+    if (res < 0 || res > INT_MAX)
+    {
+        log_error("Integer not in valid range. Expecting a positive integer with maximum size MAX_INT");
+        exit(EXIT_FAILURE);
+    }
+
+    if (endptr == str)
+    {
+        log_error("No digits found");
+        exit(EXIT_FAILURE);
+    }
+
+    const int num = res;
+
+    return num;
+}
+
+//============================================================================
+// Main entry point
+int main(int argc, char **argv)
+{
+    int timeout = 0;
+    const char* wdog_filename = "/dev/watchdog";
+
+    // Parse command line arguments
+    int current_opt;
+    int optindex;
+    static const struct option longopts[] = {
+        {"timeout", required_argument, NULL, 't'},
+        {"device", required_argument, NULL, 'w'},
+    };
+
+    while ((current_opt = getopt_long(argc, argv, "tw:", longopts, &optindex)) != -1)
+    {
+        switch (current_opt)
+        {
+            case 't':
+                timeout = numeric_string_to_int(optarg);
+                break;
+            case 'w':
+                wdog_filename = optarg;
+                break;
+            default:
+                log_warning("Only --timeout and --device should be provided. Exiting.");
+                return EXIT_FAILURE;
+        }
+    }
+
+    if (timeout <= 0)
+    {
+        log_error("--timeout must be passed and it must be a positive integer");
+        return EXIT_FAILURE;
+    }
+
+    // Set return value to EXIT_SUCCESS, will be set to EXIT_FAILURE if any of
+    // the following calls fail.
+    int ret = EXIT_SUCCESS;
+
+    const int wdog_fd = open(wdog_filename, O_WRONLY);
+    if (is_err(wdog_fd))
+    {
+        log_error("Failed to open watchdog device file");
+        ret = EXIT_FAILURE;
+        goto close;
```

Comment:
If we get here then `wdog_fd` isn't valid so it doesn't make sense to try to close it.

Reply:
8dd89775258c82300ec5d4eea7836f47c64a1584